File: sdks/java/src/main/java/com/dsai/scopeconfig/TemplateLoader.java
package com.dsai.scopeconfig;

import vn.dsai.config.v1.*;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Collectors;

import org.yaml.snakeyaml.Yaml;

/**
 * Utility class for loading configuration templates from YAML files.
 * 
 * Simply place your template files in a directory and use this class to
 * automatically load and apply them to the config service.
 * 
 * <pre>{@code
 * try (ConfigClient client = ConfigClient.fromEnvironment().build()) {
 *     TemplateLoader.loadFromDir(client, "./templates", "system");
 * }
 * }</pre>
 */
public class TemplateLoader {
    
    private static final Logger logger = Logger.getLogger(TemplateLoader.class.getName());
    
    private TemplateLoader() {
        // Utility class
    }
    
    /**
     * Load and apply all YAML templates from a directory.
     * 
     * @param client The connected ConfigClient instance
     * @param dirPath Path to the templates directory
     * @param user The user performing the action
     * @throws IOException If there's an error reading template files
     * @throws ConfigServiceException If there's an error applying templates
     */
    public static void loadFromDir(ConfigClient client, String dirPath, String user) throws IOException {
        File dir = new File(dirPath);
        
        if (!dir.exists() || !dir.isDirectory()) {
            logger.info("Templates directory " + dirPath + " does not exist, skipping template import");
            return;
        }
        
        File[] yamlFiles = dir.listFiles((d, name) -> 
            name.endsWith(".yaml") || name.endsWith(".yml"));
        
        if (yamlFiles == null || yamlFiles.length == 0) {
            logger.info("No template files found in " + dirPath);
            return;
        }
        
        logger.info("Found " + yamlFiles.length + " template file(s) to import");
        
        for (File file : yamlFiles) {
            loadAndApplyTemplateFile(client, file, user);
        }
    }
    
    /**
     * Load and apply a single YAML template file.
     * 
     * @param client The connected ConfigClient instance
     * @param filePath Path to the template file
     * @param user The user performing the action
     * @throws IOException If there's an error reading the file
     * @throws ConfigServiceException If there's an error applying the template
     */
    public static void loadFromFile(ConfigClient client, String filePath, String user) throws IOException {
        loadAndApplyTemplateFile(client, new File(filePath), user);
    }
    
    private static void loadAndApplyTemplateFile(ConfigClient client, File file, String user) throws IOException {
        Yaml yaml = new Yaml();
        Map<String, Object> data;
        
        try (FileInputStream fis = new FileInputStream(file)) {
            data = yaml.load(fis);
        }
        
        if (data == null) {
            logger.warning("Empty template file: " + file.getName());
            return;
        }
        
        // Validate required fields
        @SuppressWarnings("unchecked")
        Map<String, Object> service = (Map<String, Object>) data.get("service");
        if (service == null || !service.containsKey("id")) {
            throw new ConfigServiceException("Template file " + file.getName() + " missing 'service.id'", 
                io.grpc.Status.INVALID_ARGUMENT);
        }
        
        String serviceName = (String) service.get("id");
        String serviceLabel = (String) service.getOrDefault("label", serviceName);
        
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> groups = (List<Map<String, Object>>) data.get("groups");
        if (groups == null || groups.isEmpty()) {
            logger.warning("No groups defined in template: " + file.getName());
            return;
        }
        
        for (Map<String, Object> group : groups) {
            applyGroupTemplate(client, serviceName, serviceLabel, group, user);
            logger.info("Successfully imported template: service=" + serviceName + 
                       ", group=" + group.get("id") + " from " + file.getName());
        }
    }
    
    private static void applyGroupTemplate(
            ConfigClient client, 
            String serviceName, 
            String serviceLabel, 
            Map<String, Object> group, 
            String user) {
        
        String groupId = (String) group.getOrDefault("id", "");
        String groupLabel = (String) group.getOrDefault("label", groupId);
        String groupDescription = (String) group.getOrDefault("description", "");
        int sortOrder = ((Number) group.getOrDefault("sortOrder", 0)).intValue();
        
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> fieldMaps = (List<Map<String, Object>>) group.getOrDefault("fields", Collections.emptyList());
        
        List<ConfigFieldTemplate> fields = fieldMaps.stream()
            .map(TemplateLoader::buildFieldTemplate)
            .collect(Collectors.toList());
        
        ConfigIdentifier identifier = ConfigIdentifier.newBuilder()
            .setServiceName(serviceName)
            .setGroupId(groupId)
            .build();
        
        ConfigTemplate template = ConfigTemplate.newBuilder()
            .setIdentifier(identifier)
            .setServiceLabel(serviceLabel)
            .setGroupLabel(groupLabel)
            .setGroupDescription(groupDescription)
            .addAllFields(fields)
            .setSortOrder(sortOrder)
            .build();
        
        client.applyConfigTemplate(template, user);
    }
    
    private static ConfigFieldTemplate buildFieldTemplate(Map<String, Object> f) {
        ConfigFieldTemplate.Builder builder = ConfigFieldTemplate.newBuilder()
            .setPath((String) f.getOrDefault("path", ""))
            .setLabel((String) f.getOrDefault("label", ""))
            .setDescription((String) f.getOrDefault("description", ""))
            .setType(toFieldType((String) f.getOrDefault("type", "STRING")))
            .setDefaultValue((String) f.getOrDefault("defaultValue", ""))
            .setSortOrder(((Number) f.getOrDefault("sortOrder", 0)).intValue());
        
        @SuppressWarnings("unchecked")
        List<String> displayOn = (List<String>) f.getOrDefault("displayOn", Collections.emptyList());
        builder.addAllDisplayOn(displayOn.stream()
            .map(TemplateLoader::toScope)
            .collect(Collectors.toList()));
        
        @SuppressWarnings("unchecked")
        List<Map<String, String>> options = (List<Map<String, String>>) f.getOrDefault("options", Collections.emptyList());
        builder.addAllOptions(options.stream()
            .map(o -> ValueOption.newBuilder()
                .setValue(o.getOrDefault("value", ""))
                .setLabel(o.getOrDefault("label", o.getOrDefault("value", "")))
                .build())
            .collect(Collectors.toList()));
        
        return builder.build();
    }
    
    private static Scope toScope(String s) {
        switch (s.toUpperCase()) {
            case "SYSTEM": return Scope.SYSTEM;
            case "PROJECT": return Scope.PROJECT;
            case "STORE": return Scope.STORE;
            case "USER": return Scope.USER;
            default: return Scope.SCOPE_UNSPECIFIED;
        }
    }
    
    private static FieldType toFieldType(String t) {
        switch (t.toUpperCase()) {
            case "STRING": return FieldType.STRING;
            case "INT": return FieldType.INT;
            case "FLOAT": return FieldType.FLOAT;
            case "BOOLEAN": return FieldType.BOOLEAN;
            case "JSON": return FieldType.JSON;
            case "ARRAY_STRING": return FieldType.ARRAY_STRING;
            case "SECRET": return FieldType.SECRET;
            default: return FieldType.STRING;
        }
    }
}

File: sdks/go/template_loader.go
package scopeconfig

import (
	"context"
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"strings"

	configv1 "github.com/digitalsolutionsai/scope-config-service/sdks/go/gen/config/v1"
	"gopkg.in/yaml.v2"
)

// YamlValueOption represents a value option in YAML format.
type YamlValueOption struct {
	Value string `yaml:"value"`
	Label string `yaml:"label"`
}

// YamlFieldTemplate represents a field template in YAML format.
type YamlFieldTemplate struct {
	Path         string            `yaml:"path"`
	Label        string            `yaml:"label"`
	Description  string            `yaml:"description"`
	Type         string            `yaml:"type"`
	DefaultValue string            `yaml:"defaultValue"`
	DisplayOn    []string          `yaml:"displayOn"`
	Options      []YamlValueOption `yaml:"options"`
	SortOrder    int32             `yaml:"sortOrder"`
}

// YamlGroup represents a configuration group in YAML format.
type YamlGroup struct {
	ID          string              `yaml:"id"`
	Label       string              `yaml:"label"`
	Description string              `yaml:"description"`
	Fields      []YamlFieldTemplate `yaml:"fields"`
	SortOrder   int32               `yaml:"sortOrder"`
}

// YamlTemplate represents the complete template structure in YAML format.
type YamlTemplate struct {
	Service struct {
		Name  string `yaml:"id"`
		Label string `yaml:"label"`
	} `yaml:"service"`
	Groups []YamlGroup `yaml:"groups"`
}

// LoadTemplatesFromDir loads all YAML templates from the specified directory
// and applies them using the client.
//
// Example:
//
//	err := client.LoadTemplatesFromDir(ctx, "./templates", "system")
func (c *Client) LoadTemplatesFromDir(ctx context.Context, dir string, user string) error {
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		log.Printf("Templates directory %s does not exist, skipping template import", dir)
		return nil
	}

	var templateFiles []string
	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if ext == ".yml" || ext == ".yaml" {
			templateFiles = append(templateFiles, path)
		}
		return nil
	})
	if err != nil {
		return fmt.Errorf("failed to walk templates directory: %w", err)
	}

	if len(templateFiles) == 0 {
		log.Printf("No template files found in %s", dir)
		return nil
	}

	log.Printf("Found %d template file(s) to import", len(templateFiles))

	for _, file := range templateFiles {
		if err := c.loadAndApplyFile(ctx, file, user); err != nil {
			return fmt.Errorf("failed to apply template from %s: %w", file, err)
		}
	}

	return nil
}

func (c *Client) loadAndApplyFile(ctx context.Context, filePath string, user string) error {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return fmt.Errorf("failed to read file: %w", err)
	}

	var yamlTemplate YamlTemplate
	if err := yaml.Unmarshal(data, &yamlTemplate); err != nil {
		return fmt.Errorf("failed to parse YAML: %w", err)
	}

	if yamlTemplate.Service.Name == "" {
		return fmt.Errorf("service.id is required in template file")
	}

	for _, group := range yamlTemplate.Groups {
		if err := c.applyGroup(ctx, &yamlTemplate, &group, user); err != nil {
			return fmt.Errorf("failed to apply group %s: %w", group.ID, err)
		}
		log.Printf("Successfully imported template: service=%s, group=%s from %s",
			yamlTemplate.Service.Name, group.ID, filepath.Base(filePath))
	}

	return nil
}

func (c *Client) applyGroup(ctx context.Context, yamlTemplate *YamlTemplate, group *YamlGroup, user string) error {
	fields := make([]*configv1.ConfigFieldTemplate, len(group.Fields))
	for i, f := range group.Fields {
		displayOn := make([]configv1.Scope, len(f.DisplayOn))
		for j, d := range f.DisplayOn {
			displayOn[j] = toScope(d)
		}

		options := make([]*configv1.ValueOption, len(f.Options))
		for j, o := range f.Options {
			options[j] = &configv1.ValueOption{Value: o.Value, Label: o.Label}
		}

		fields[i] = &configv1.ConfigFieldTemplate{
			Path:         f.Path,
			Label:        f.Label,
			Description:  f.Description,
			Type:         toFieldType(f.Type),
			DefaultValue: f.DefaultValue,
			DisplayOn:    displayOn,
			Options:      options,
			SortOrder:    f.SortOrder,
		}
	}

	template := &configv1.ConfigTemplate{
		Identifier: &configv1.ConfigIdentifier{
			ServiceName: yamlTemplate.Service.Name,
			GroupId:     group.ID,
		},
		ServiceLabel:     yamlTemplate.Service.Label,
		GroupLabel:       group.Label,
		GroupDescription: group.Description,
		Fields:           fields,
		SortOrder:        group.SortOrder,
	}

	_, err := c.ApplyConfigTemplate(ctx, template, user)
	return err
}

func toFieldType(s string) configv1.FieldType {
	switch s {
	case "STRING":
		return configv1.FieldType_STRING
	case "INT":
		return configv1.FieldType_INT
	case "FLOAT":
		return configv1.FieldType_FLOAT
	case "BOOLEAN":
		return configv1.FieldType_BOOLEAN
	case "JSON":
		return configv1.FieldType_JSON
	case "ARRAY_STRING":
		return configv1.FieldType_ARRAY_STRING
	case "SECRET":
		return configv1.FieldType_SECRET
	default:
		return configv1.FieldType_FIELD_TYPE_UNSPECIFIED
	}
}

func toScope(s string) configv1.Scope {
	switch s {
	case "SYSTEM":
		return configv1.Scope_SYSTEM
	case "PROJECT":
		return configv1.Scope_PROJECT
	case "STORE":
		return configv1.Scope_STORE
	case "USER":
		return configv1.Scope_USER
	default:
		return configv1.Scope_SCOPE_UNSPECIFIED
	}
}

// LoadTemplateFromBytes loads a single template from YAML bytes and applies it.
//
// Example:
//
//	yamlData := []byte(`
//	service:
//	  id: "my-service"
//	  label: "My Service"
//	groups:
//	  - id: "settings"
//	    label: "Settings"
//	    fields:
//	      - path: "log.level"
//	        label: "Log Level"
//	        type: "STRING"
//	        defaultValue: "INFO"
//	`)
//	err := client.LoadTemplateFromBytes(ctx, yamlData, "system")
func (c *Client) LoadTemplateFromBytes(ctx context.Context, data []byte, user string) error {
	var yamlTemplate YamlTemplate
	if err := yaml.Unmarshal(data, &yamlTemplate); err != nil {
		return fmt.Errorf("failed to parse YAML: %w", err)
	}

	if yamlTemplate.Service.Name == "" {
		return fmt.Errorf("service.id is required in template")
	}

	for _, group := range yamlTemplate.Groups {
		if err := c.applyGroup(ctx, &yamlTemplate, &group, user); err != nil {
			return fmt.Errorf("failed to apply group %s: %w", group.ID, err)
		}
	}

	return nil
}
